File: machine/ts/render/host_overlay/pipeline.ts
import type { HostMenuPipelineState, HostOverlayPipelineState, RenderGraphPassContext } from '../backend/backend';
import { consumeHostMenuFrame, consumeOverlayFrame } from './overlay_queue';

const EMPTY_HOST_2D_KINDS = [];
const EMPTY_HOST_2D_REFS = [];

export function createHostOverlayState(): HostOverlayPipelineState {
	return {
		width: 0,
		height: 0,
		overlayWidth: 0,
		overlayHeight: 0,
		time: 0,
		delta: 0,
		commandKinds: EMPTY_HOST_2D_KINDS,
		commandRefs: EMPTY_HOST_2D_REFS,
		commandCount: 0,
	};
}

export function writeHostOverlayState(ctx: RenderGraphPassContext, state: HostOverlayPipelineState): void {
	const frame = consumeOverlayFrame();
	state.width = frame.renderWidth;
	state.height = frame.renderHeight;
	state.overlayWidth = frame.logicalWidth;
	state.overlayHeight = frame.logicalHeight;
	state.time = ctx.time;
	state.delta = ctx.delta;
	state.commandKinds = frame.commandKinds;
	state.commandRefs = frame.commandRefs;
	state.commandCount = frame.commandCount;
}

export function createHostMenuState(): HostMenuPipelineState {
	return {
		width: 0,
		height: 0,
		overlayWidth: 0,
		overlayHeight: 0,
		time: 0,
		delta: 0,
		commandKinds: EMPTY_HOST_2D_KINDS,
		commandRefs: EMPTY_HOST_2D_REFS,
		commandCount: 0,
	};
}

export function writeHostMenuState(ctx: RenderGraphPassContext, state: HostMenuPipelineState): void {
	const presenter = ctx.presenter;
	const frame = consumeHostMenuFrame();
	state.width = presenter.offscreenCanvasSize.x;
	state.height = presenter.offscreenCanvasSize.y;
	state.overlayWidth = presenter.viewportSize.x;
	state.overlayHeight = presenter.viewportSize.y;
	state.time = ctx.time;
	state.delta = ctx.delta;
	state.commandKinds = frame.commandKinds;
	state.commandRefs = frame.commandRefs;
	state.commandCount = frame.commandCount;
}

File: machine/cpp/render/host_overlay/pipeline.cpp
#include "render/host_overlay/pipeline.h"

#include "render/video_presenter.h"

namespace bmsx {
void writeHostOverlayPassState(
	const RenderPassDef::RenderGraphPassContext& ctx,
	RenderPassStateStorage& stateStorage
) {
	HostOverlayPipelineState& state = stateStorage.hostOverlay;
	VideoPresenter& presenter = *ctx.presenter;
	state.time = ctx.time;
	state.delta = ctx.delta;
	const HostOverlayFrame frame = presenter.hostOverlayQueue.consumeOverlayFrame();
	state.width = frame.renderWidth;
	state.height = frame.renderHeight;
	state.overlayWidth = frame.logicalWidth;
	state.overlayHeight = frame.logicalHeight;
	state.commandKinds = frame.commandKinds;
	state.commandRefs = frame.commandRefs;
	state.commandCount = frame.commandCount;
}

void writeHostMenuPassState(
	const RenderPassDef::RenderGraphPassContext& ctx,
	RenderPassStateStorage& stateStorage
) {
	HostMenuPipelineState& state = stateStorage.hostMenu;
	VideoPresenter& presenter = *ctx.presenter;
	state.width = static_cast<i32>(presenter.offscreenCanvasSize.x);
	state.height = static_cast<i32>(presenter.offscreenCanvasSize.y);
	state.overlayWidth = static_cast<i32>(presenter.viewportSize.x);
	state.overlayHeight = static_cast<i32>(presenter.viewportSize.y);
	const HostMenuFrame frame = presenter.hostOverlayQueue.consumeHostMenuFrame();
	state.commandKinds = frame.commandKinds;
	state.commandRefs = frame.commandRefs;
	state.commandCount = frame.commandCount;
}

bool shouldExecuteHostOverlayPass(VideoPresenter* presenter, void*) {
	return presenter->hostOverlayQueue.hasPendingOverlayFrame();
}

bool shouldExecuteHostMenuPass(VideoPresenter* presenter, void*) {
	return presenter->hostOverlayQueue.hasPendingHostMenuFrame();
}

} // namespace bmsx
